### ISSUE TITLE
Implicit binding of name to reference bug

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2228]
Migrated from [rt.perl.org#3544](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3544) (status was 'rejected')

Searchable as RT3544$

**Comment by p5pRT** [comment]
### From consulting@altnet.net

#### Created by consulting@altnet.net



I'm not sure if this is a bug in the implementation or the definition of the
language\, but it is definitely wrong\.  Accessing a reference that does not
exist by using a scalar variable will as a side effect assign to the scalar
being used\.  For example\, this&#8203;:

my $foo = undef;
print "\\$foo is now $foo\\n";
for \(@&#8203;$foo\) \{ print "\\t==>$\_\\n"; \}
print "\\$foo is now $foo\\n";

prints out something like this&#8203;:

$foo is now 
$foo is now ARRAY\(0x804b490\)

Notice that nothing explictly assigned to $foo\.  It happened as a side
effect of using an undefined value as an ARRAY ref\.  This similar code
fragment gives radically different results&#8203;:

use strict;
my $foo = "Kermit\, the frog";
print "\\$foo is now $foo\\n";
for \(@&#8203;$foo\) \{ print "\\t==>$\_\\n"; \}
print "\\$foo is now $foo\\n";

prints out this&#8203;:

$foo is now Kermit\, the frog
Can't use string \("Kermit\, the frog"\) as an ARRAY ref while "strict refs" in use at \-e line 5\.

Without the "use strict"\, it prints out&#8203;:

$foo is now Kermit\, the frog
$foo is now Kermit\, the frog

The first code fragment behaves exactly the same with and without the "use
strict"\.  This behavior is certainly inconsistent and counter\-intuitive\, and I
am not aware of any requirement of the language to behave this way\.  I have
also seen this behavior in version 5\.004\_01 and 5\.005\_03\.

Suggested fixes&#8203;:
1\)  "strict refs" should\, in fact\, enforce that an actual reference exists
when $foo is used as a reference\.
2\)  The implicitly created reference in the case where "strict refs" is not
in force should be just as implicitly discarded when it goes out of scope\,
not implicitly bound to a name\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by root at Fri Apr  7 11:47:37 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.14-5.0, archname=i686-linux
    uname='linux jschneid-lin 2.2.14-5.0 #1 tue mar 7 21:07:39 est 2000 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing'
    ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i686-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl/5.005
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/jschneid
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/bin:/usr/local/bin:/usr/X11R6/bin:/home/jschneid/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @tamias

On Mon\, Jul 17\, 2000 at 02&#8203;:09&#8203;:04PM \-0700\, consulting@&#8203;altnet\.net wrote&#8203;:

> I'm not sure if this is a bug in the implementation or the definition of the
> language\, but it is definitely wrong\.  Accessing a reference that does not
> exist by using a scalar variable will as a side effect assign to the scalar
> being used\.  For example\, this&#8203;:
> 
> my $foo = undef;
> print "\\$foo is now $foo\\n";
> for \(@&#8203;$foo\) \{ print "\\t==>$\_\\n"; \}
> print "\\$foo is now $foo\\n";
> 
> prints out something like this&#8203;:
> 
> $foo is now 
> $foo is now ARRAY\(0x804b490\)

This is not a bug\.  It is autovivification\.  RTFM; perlref in this case\.

Ronald

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Jul 17\, consulting@&#8203;altnet\.net said&#8203;:

>my $foo = undef;
>print "\\$foo is now $foo\\n";
>for \(@&#8203;$foo\) \{ print "\\t==>$\_\\n"; \}
>print "\\$foo is now $foo\\n";

From the perlref documentation&#8203;:

&nbsp;             $array\[$x\]\->\{"foo"\}\->\[0\] = "January";

&nbsp;         This is one of the cases we mentioned earlier in which
&nbsp;         references could spring into existence when in an lvalue
&nbsp;         context\.  Before this statement\, $array\[$x\] may have
&nbsp;         been undefined\.  If so\, it's automatically defined with
&nbsp;         a hash reference so that we can look up \{"foo"\} in it\.
&nbsp;         Likewise $array\[$x\]\->\{"foo"\} will automatically get
&nbsp;         defined with an array reference so that we can look up
&nbsp;         \[0\] in it\.  This process is called autovivification\.

Basically\, an undefined value can be turned into a reference implicitly as
you have just shown\.

There is a SEVERE difference between&#8203;:

&nbsp;  undef $x;    @&#8203;$x = \(1\,2\,3\);

and

&nbsp;  $x = "foo";  @&#8203;$x = \(1\,2\,3\);

The former autovivifies an array reference in $x\.  The latter creates the
array @&#8203;foo through symbolic referencing\.

I am puzzled\, though\, about perldiag's message of&#8203;:

&nbsp;     Can't use an undefined value as %s reference
&nbsp;         \(F\) A value used as either a hard reference or a
&nbsp;         symbolic reference must be a defined value\.  This helps
&nbsp;         to delurk some insidious errors\.

What causes this to be raised?

**Comment by p5pRT** [comment]
### From nobull@mail.com

> I am puzzled\, though\, about perldiag's message of&#8203;:
> 
>      Can't use an undefined value as %s reference
>          \(F\) A value used as either a hard reference or a
>          symbolic reference must be a defined value\.  This helps
>          to delurk some insidious errors\.
> 
> What causes this to be raised?

Autovivification does not apply in a simple rvalue context\.  \(But perl
errs on the side of allowing autovivification\)\.

my $x;
print $$x;

Can't use an undefined value as a SCALAR reference

my $x;
print $\{\\$$x\};
print $x;

Use of uninitialized value in print
SCALAR\(0x224f18\)

**Comment by p5pRT** [comment]
### From nobull@mail.com

> I am puzzled\, though\, about perldiag's message of&#8203;:
> 
>      Can't use an undefined value as %s reference
>          \(F\) A value used as either a hard reference or a
>          symbolic reference must be a defined value\.  This helps
>          to delurk some insidious errors\.
> 
> What causes this to be raised?

Autovivification does not apply in a simple rvalue context\.  \(But perl
errs on the side of allowing autovivification\)\.

my $x;
print $$x;

Can't use an undefined value as a SCALAR reference

my $x;
print $\{\\$$x\};
print $x;

Use of uninitialized value in print
SCALAR\(0x224f18\)

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'rejected'
